src/zkscript/bilinear_pairings/model/triple_miller_loop.py

```diff
@@ -77,1415 +547,129 @@ def triple_miller_loop(
-                out += roll(position=3 * N_POINTS_TWIST + stack_length_added - 1, n_elements=N_POINTS_TWIST)  # Roll T2
-                stack_length_added += N_POINTS_TWIST
-                out += point_doubling_twisted_curve(
-                    take_modulo=take_modulo_T,
-                    verify_gradient=True,
-                    check_constant=False,
-                    clean_constant=False,
-                    positive_modulo=False,
-                )  # Compute 2*T2
-                # After this, the stack is: P1 P2 P3 Q1 Q2 Q3 -Q1 -Q2 -Q3 (2*T1) (2*T2) (2*T3),
-                # altstack = [(f_i^2 * t_1 * t_2 * t_3)]
-                stack_length_added = 0
-                out += roll(
-                    position=9 * N_POINTS_TWIST + 3 * N_POINTS_CURVE + EXTENSION_DEGREE + stack_length_added - 1,
-                    n_elements=EXTENSION_DEGREE,
-                )  # Roll lambda_(2*T3)
-                stack_length_added += EXTENSION_DEGREE
-                out += roll(position=3 * N_POINTS_TWIST + stack_length_added - 1, n_elements=N_POINTS_TWIST)  # Roll T3
-                stack_length_added += N_POINTS_TWIST
-                out += point_doubling_twisted_curve(
-                    take_modulo=take_modulo_T,
-                    verify_gradient=True,
-                    check_constant=False,
-                    clean_constant=clean_final,
+            if i != len(self.exp_miller_loop) - 2:
+                # stack in:  [P1, P2, P3, Q1, Q2, Q3, T1, T2, T3, f_i]
+                # stack out: [P1, P2, P3, Q1, Q2, Q3, T1, T2, T3, f_i^2]
+                out += self.miller_loop_output_square(take_modulo=False, check_constant=False, clean_constant=False)
+            if self.exp_miller_loop[i] == 0:
+                # stack in:  [gradient_(2*T1), gradient_(2*T2), gradient_(2*T3), P1, P2, P3, Q1, Q2, Q3, T1, T2, T3,
+                #               {f_i^2}]
+                # stack out: [P1, P2, P3, Q1, Q2, Q3, (2*T1), (2*T2), (2*T3),
+                #               {f_i^2} * (ev_(l_(T1,T1))(P1) * ev_(l_(T2,T2))(P2) * ev_(l_(T3,T3))(P3)]
+                out += self.__one_step_without_addition(
+                    i=i,
+                    take_modulo=[take_modulo_miller_loop_output, take_modulo_point_multiplication],
                     positive_modulo=positive_modulo_i,
-                )  # Compute 2*T3
-                # After this, the stack is: P1 P2 P3 Q1 Q2 Q3 -Q1 -Q2 -Q3 (2*T1) (2*T2) (2*T3) (f_i^2 * t_1 * t_2 * t_3)
-                out += Script.parse_string(" ".join(["OP_FROMALTSTACK"] * N_ELEMENTS_MILLER_OUTPUT))
-            else:
-                # After this, the stack is: lambda_(2*T1) lambda_(2*T2) lambda_(2*T3) P1 P2 P3 Q1 Q2 Q3 -Q1 -Q2 -Q3
-                # T1 T2 T3 f_i^2
-                out += miller_loop_output_square(
-                    take_modulo=False, positive_modulo=False, check_constant=False, clean_constant=False
+                    clean_constant=clean_constant,
```

Comment:
I would either have a variable clean_constant_i = ... or change the behaviour of positive modulo so they are consistent.